src/doc_table.c

```diff
@@ -9,24 +9,76 @@
 #include "rmalloc.h"
 #include "spec.h"
 
+#define MAX_SIZE 1000000
+
 /* Creates a new DocTable with a given capacity */
 DocTable NewDocTable(size_t cap) {
   return (DocTable){.size = 1,
                     .cap = cap,
                     .maxDocId = 0,
                     .memsize = 0,
                     .sortablesSize = 0,
-                    .docs = rm_calloc(cap, sizeof(RSDocumentMetadata)),
+                    .maxSize = MAX_SIZE,
+                    .buckets = rm_calloc(cap, sizeof(DMDChain)),
                     .dim = NewDocIdMap()};
 }
 
+static inline uint32_t DocTable_GetBucket(DocTable *t, t_docId docId) {
+  return docId < t->maxSize ? docId : docId % t->maxSize;
+}
+
+static inline int DocTable_ValidateDocId(DocTable *t, t_docId docId) {
+  return docId != 0 && docId <= t->maxDocId;
+}
+
 /* Get the metadata for a doc Id from the DocTable.
  *  If docId is not inside the table, we return NULL */
+
+int DMDChain_IsEmpty(DMDChain *dmdChain) {
+  return !dmdChain->last;
+}
+
 inline RSDocumentMetadata *DocTable_Get(DocTable *t, t_docId docId) {
-  if (docId == 0 || docId > t->maxDocId) {
+  if (!DocTable_ValidateDocId(t, docId)) {
     return NULL;
   }
-  return &t->docs[docId];
+  uint32_t bucketIndex = DocTable_GetBucket(t, docId);
+  DMDChain *dmdChain = &t->buckets[bucketIndex];
+  RSDocumentMetadata *currDmd = dmdChain->first;
+  while (currDmd) {
+    if (currDmd->id == docId) {
+      return currDmd;
+    }
+    currDmd = currDmd->next;
+  }
+  return NULL;
+}
+
+static inline void DocTable_Set(DocTable *t, t_docId docId, RSDocumentMetadata dmd) {
```

Comment:
Please pass dmd by pointer, not by value. I know the compiler could technically optimize this out, but let's not rely on that (besides, I'm unsure).